### PR TITLE
Support implicit/explicit var with inferred lambda types

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/58404

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs [hunk 1]
```diff
@@ -15,6 +15,7 @@
 using Microsoft.CodeAnalysis.Simplification;
 using Microsoft.CodeAnalysis.Operations;
 using Microsoft.CodeAnalysis.CSharp.Simplification;
+using System;
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs:18]: >System [](http://example.com/codeflow?start=6&length=6)

nit: sort

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs [hunk 2]
```diff
@@ -333,6 +331,19 @@ protected override bool AssignmentSupportsStylePreference(
...
 
+        // This also applies to a lambda assigned to a variable.  This will have no conversion, but can be converted as
+        // long as the type is Func<> or Action<> as that's what the language will infer here.
+        if (!conversion.Exists && expression is LambdaExpressionSyntax && semanticModel.Compilation.LanguageVersion() >= LanguageVersion.CSharp10)
+        {
+            var initializerType = semanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            return declaredType.Equals(initializerType) &&
+                declaredType is
+                {
+                    Name: nameof(Func<int>) or nameof(Action<int>),
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs:342]: >int [](http://example.com/codeflow?start=38&length=3)

intentional <int>?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs]: Done.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs:341]: Yes.  The language doesn't allow leaving out a type-parameter in nameof.  I championed and implemented the feature to allow eliding this.  But that comes in C# 14, which we are not on yet.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**